net/nettest/conn.go

```diff
@@ -40,24 +32,38 @@ func NewConn(name string, maxBuf int) (Conn, Conn) {
 	return &connHalf{r: r, w: w}, &connHalf{r: w, w: r}
 }
 
+type connAddr string
+
+func (a connAddr) Network() string { return "mem" }
+func (a connAddr) String() string  { return string(a) }
+
 type connHalf struct {
 	r, w *Pipe
 }
 
+func (c *connHalf) LocalAddr() net.Addr {
+	return connAddr(c.r.name)
+}
+
+func (c *connHalf) RemoteAddr() net.Addr {
+	return connAddr(c.w.name)
+}
+
 func (c *connHalf) Read(b []byte) (n int, err error) {
 	return c.r.Read(b)
 }
 func (c *connHalf) Write(b []byte) (n int, err error) {
 	return c.w.Write(b)
 }
+
 func (c *connHalf) Close() error {
-	err1 := c.r.Close()
-	err2 := c.w.Close()
-	if err1 != nil {
-		return err1
+	if err := c.w.Close(); err != nil {
+		return err
 	}
-	return err2
+	c.r.CloseWrite()
```

Comment:
what's with this part of the change? Why did this Close change to a real Close in one direction but a CloseWrite in the other?

Reply:
IIUC in a client/server interaction, connections are usually closed from a single side. The other side doesn't see anything until the read buffer is drained. With the initial implementation we would see error when the other side closed the connection.
However I can simplify this just by returning nil from Close if the pipe has already been closed.